### ISSUE TITLE
Create new storage engine

### ISSUE THREAD

**Issue by pauldix** [influxdb/influxdb#4086]
The why is littered over the last two years of issues. The how is my well guarded secret. This is mainly to track things so people can see when it's done.

I'll put together a very detailed writeup when this is done, but in the meantime here's what it should do:

* Shouldn't require data migration for users upgrading from 0.9.x
* Hot backups
* Compression significantly better than what is available in 0.9.4
* Write throughput that doesn't spike IOPS and can keep up with the WAL
* Reads shouldn't block writes
* Writes shouldn't block reads

**Comment by jhorwit2** [comment]
:+1:

**Comment by Jhors2** [comment]
Paul, can you at all comment about read performance as well?  Obviously the non blocking architecture should make read times more consistent.  Just curious what preliminary benchmarks of the storage engine are looking like.

**Comment by pauldix** [comment]
@Jhors2: It's hard to say at this point. We're still working out some stuff on compression. I benchmarked something and it was slightly faster than 0.9.4 (bz1), but we really won't know for sure until later.

The storage engine itself should be fairly fast if your working set fits in memory. We'll be improving query performance through updates in the query engine.

**Comment by kev009** [comment]
I'd like to see knobs so application compression can be disabled.

**Comment by Jhors2** [comment]
+1, compaction knobs would be nice.  This would help you make the decision of "I don't care if it uses more disk, I need as much query/write performance as possible"

**Comment by pauldix** [comment]
Having a knob to disable compression may come later, but it's not likely to make it in the first version.

Also, if compression makes it so that your working query set fits in memory, it could actually have a positive impact on query performance. We use memory mapped files so that's a possibility.

**Comment by kev009** [comment]
You're having to work around legacy file systems.  Compression, snapshots, high availability, intelligent mgmt of working set can be handily had without much code by using ZFS.  If you haven't, reading up on Circonus's snowth may be enlightening because it's the only production quality metrics system I'm aware of at the moment.

**Comment by victorhooi** [comment]
I'm just curious if there's a timeframe on when the new storage engine might hit nightlies?

Super excited to try it!

Also - would it resolve the (possible) snappy compression issues in #4135?

**Comment by syepes** [comment]
+1 This is a must have as in the current stat its completely useless, really hope the new SE resolves the memory and write throughput issues.
Just Off-topic, did you guys consider using a Non GC language like Rust?

**Comment by pauldix** [comment]
@victorhooi: We're working feverishly on it, we'll announce as soon as it hits the nightly. It will almost certainly resolve #4135. Will almost certainly be within the next two weeks (but not before next Friday).

@syepes: useless for your workload perhaps, but there are plenty of people using it already with the old engine. But point taken, we're focused very heavily on this. It's the number one priority.

We didn't consider Rust because 2 years ago when we started this project it wasn't close to being ready. We really only debated at the time between C, C++, or Go. The latter obviously won :)

**Comment by victorhooi** [comment]
I saw there's a new TSM engine that just got merged? =)

https://github.com/influxdb/influxdb/commit/f041939a1cbac85a4fc109ae1112ade0fb26b77e

Should I be keeping my eyes open for an announcement on the blog, talking about this new awesomeness? =)

**Comment by pauldix** [comment]
haha yes, should have details up in about 12 hours.

**Comment by pauldix** [comment]
This is ready for testing. More details:
Blog:
https://influxdb.com/blog/2015/10/07/the_new_influxdb_storage_engine_a_time_structured_merge_tree.html
Detailed writeup:
https://influxdb.com/docs/v0.9/concepts/storage_engine.html
Installation instructions (you have to enable it):
https://influxdb.com/docs/v0.9/introduction/tsm_installation.html

The PR was #4308

**Comment by fg2it** [comment]
@pauldix thanks for the details and this new engine.
A question : if I understood correclty, you don't expect integer values to be correlated, contrary to float values which you expect to be close if consecutive, is it the result of some benchmarking ? Naively, as time serie values, I would have expected integer values to show the same kind of behavior as float. So I am surprised you don't perform delta-encoding before the zig-zag one. Moreover, I would expect that delta-encoding don't degrade performances on uncorrelated values.

**Comment by myurko2** [comment]
+1 to @fg2it 's comment. For my envisioned use of integer values, a delta encoding makes the most sense.

**Comment by jwilder** [comment]
@fg2it @myurko2 delta-encoding for int64s might be used in the future.  There is room for additional encodings and improvements.

**Comment by RobMcZag** [comment]
My experience with HP Vertica is that having some sensible default with the ability to chose from a few alternatives the right encoding for every field when you need it is a real benefit.
Only at design time (and sometimes only after checking real data) you have a clear idea of the contents, if you are going to order on that column or not and how to optimize the storage. Once you start with one encoding you have to recreate the table to change it.

As an example Run Length Encoding and Dictionary are very effective with string categories (especially with long names), but you have to know if you are sorting on that column and the cardinality of the values. 

I understand that in InfluxDB we do not declare a schema or fields type, but it nevertheless gets decided at the first insert with such a field. Would be practical being able to have a dummy insert just to have the field created and set with the desired default?

**Comment by otoolep** [comment]
>I understand that in InfluxDB we do not declare a schema or fields type, but it nevertheless gets decided at the first insert with such a field. Would be practical being able to have a dummy insert just to have the field created and set with the desired default?

This will work, however there is a caveat. Field types are only on a per-shard basis (under the covers, this is not something exposed to users) so it would require you perform that dummy insert as the first event in each shard, but when that shard is created is not visible, nor controllable by you (not directly anyway).

One possibility, of course, is for the system to use a schema if you supply it. But if a schema is not supplied, the types would be inferred (which is what happens today). If you think this would be useful, the best thing to do it to open a new ticket, indicating that it is a feature request. That way others in the community can weigh in
